crates/oxide/src/lib.rs

```diff
@@ -65,24 +74,89 @@ pub fn clear_cache() {
     cache.clear();
 }
 
-pub fn scan_dir(opts: ScanOptions) -> ScanResult {
-    init_tracing();
+#[derive(Debug, Clone, Default)]
+pub struct Scanner {
+    /// Auto content configuration
+    auto_content: Option<AutoContent>,
+
+    /// Glob sources
+    sources: Option<Vec<GlobEntry>>,
+
+    /// All files that we have to scan
+    files: Vec<PathBuf>,
+
+    /// All generated globs
+    globs: Vec<GlobEntry>,
+
+    /// All found candidates in the files
+    candidates: Vec<String>,
+}
 
-    let (mut files, mut globs) = match opts.base {
-        Some(base) => {
-            // Only enable auto content detection when `base` is provided.
-            let base = Path::new(&base);
-            let (files, dirs) = resolve_files(base);
-            let globs = resolve_globs(base, dirs);
+impl Scanner {
+    pub fn new(auto_content: Option<AutoContent>, sources: Option<Vec<GlobEntry>>) -> Self {
+        init_tracing();
 
-            (files, globs)
+        let mut scanner = Self {
+            auto_content,
+            sources,
+            ..Default::default()
+        };
+
+        scanner.scan();
```

Comment:
It feels super weird to me that `::new()` has the side effect of touching the filesystem. I think I'd prefer that the users for `Scanner` call `scan` instead of relying on `::new()` basically not returning until its done.

Reply:
Yep I agree, reasons I still did it is that when using it, we basically always did this:

```ts
let scanner = new Scanner(…);
scanner.scan();
```

I think in a perfect world you can do something like this:
```ts
let scanner = new Scanner(…);

// Setup watchers, so that they are already "listening" for changes before we perform the first scan. Prevents a potential race condition where a file changed between setting up the scanner and creating the watchers.
await setupWatchers();

// Perform initial scan
scanner.scan();
```

However, in this case, in order to setup the watchers, we need to resolve the globs (which touches the file system).